pkg/manager/internal.go

```diff
@@ -557,7 +561,12 @@ func (cm *controllerManager) waitForRunnableToEnd(shutdownCancel context.CancelF
 	// Cancel leader election only after we waited. It will os.Exit() the app for safety.
 	defer func() {
 		if cm.leaderElectionCancel != nil {
+			// After asking the context to be cancelled, make sure
+			// we wait for the leader stopped channel to be closed, otherwise
+			// we might encounter race conditions between this code
+			// and the event recorder, which is used within leader election code.
 			cm.leaderElectionCancel()
+			<-cm.leaderElectionStopped
```

Comment:
You mean checking the return `err` and only if it's not nil, cancel leader election? The error comes from `shutdownCtx`, if we don't cancel leader election, I'd assume that it'll timeout after a bit and the main.go should exit, right?

Reply:
Trying to double check the thinking, I'd also like to capture in a comment around the err check

Reply:
only if its nil, cancel leader election, if its not-nil, that means something is still running and its unsafe to cancel leader election because that might result in the thing still running after we are not leader anymore.

This should result in the binary exiting which exists all pending routines and the leader lease of us will just time out.